cloudferrylib/os/identity/keystone.py

```diff
@@ -474,7 +477,57 @@ def _get_user_tenants_roles(self, tenant_list=None, user_list=None):
             tenant_list = []
         if user_list is None:
             user_list = []
-        user_tenants_roles = {}
+        if not self.config.identity.optimize_user_role_fetch:
+            user_tenants_roles = self._get_user_tenants_roles_by_api(tenant_list,
+                                                                     user_list)
+        else:
+            user_tenants_roles = self._get_user_tenants_roles_by_db(tenant_list,
+                                                                    user_list)
+        return user_tenants_roles
+
+    def _get_user_roles_cached(self):
+        all_roles = {}
+        if self.config.identity.optimize_user_role_fetch:
```

Comment:
Also code which is not covered with tests is by definition not maintainable, so please refactor.